### DISCUSSION TITLE
Pressure as initial condition for sink

### DISCUSSION THREAD

**Discussion by dad616610** [e2nIEE/pandapipes#621] [Q&A]
Hello!

I'm trying to set pressure as initial condition for sink (here by sink I mean network output) instead of mass flow. The desired sink pressure is 1 bar (might need to set a different pressure as well, but 1 bar will be sufficient for now)

I've tried using External Grid (because I can specify pressure as an initial condition here), but the solution doesn't converge. I've also seen this question: https://github.com/e2nIEE/pandapipes/discussions/410. I tried the example provided as is, but I get the following error:
```
Traceback (most recent call last):
  File "<..>/test_control.py", line 27, in <module>
    run_control(net)  # --> pipeflow calculation that applies the CharacteristicControl during the simulation
  File "<..>/pandapipes/src/pandapipes/control/run_control.py", line 28, in run_control
    run_control_pandapower(net, ctrl_variables=ctrl_variables, max_iter=max_iter, **kwargs)
  File "<..>/.venv/lib/python3.10/site-packages/pandapower/control/run_control.py", line 303, in run_control
    control_implementation(net, controller_order, ctrl_variables, max_iter, **kwargs)
  File "<..>/.venv/lib/python3.10/site-packages/pandapower/control/run_control.py", line 214, in control_implementation
    ctrl_converged = _control_step(levelorder, run_count)
  File "<..>/.venv/lib/python3.10/site-packages/pandapower/control/run_control.py", line 238, in _control_step
    if not ctrl.is_converged(net):
  File "<..>/.venv/lib/python3.10/site-packages/pandapower/control/controller/characteristic_control.py", line 82, in is_converged
    self.values = net.characteristic.object.at[self.characteristic_index](input_values)
  File "<..>/.venv/lib/python3.10/site-packages/pandapower/auxiliary.py", line 188, in __getattr__
    raise AttributeError(
AttributeError: 'pandapipesNet' instance has no attribute 'characteristic'

```

So, how do I specify the pressure for a sink?

**Comment by dlohmeier** [comment]
Dear @dad616610 ,

that is an interesting issue, but I think I would need to have a look at the model you would like to implement (i.e. the network) to give you really helpful advise. A few notes here:
- If you would like implement a fixed pressure at a specific node, an ext_grid is definitely the way to go. If the simulation does not converge, it probably means that the setup is not enabling a proper solution.
- The issue you mentioned has a slightly different objective: there the mass flow at a sink is not constant, but depends on the pressure at the node. It does not determine the pressure.
- 
If I understand you correctly, you really want a fixed pressure and the mass flow should be set accordingly. In that case, an ext_grid is the way to go. If you need help finding the right setup, please provide some code that helps to understand the problem.
Kind regards!

**Comment by dad616610** [comment]
Hello, @dlohmeier,

Thank you for your reply. Sorry it took me so long to reply!

I was a bit misleading when I said that the solution

> didn't converge

Actually, it was not the solution, but the Colebrook-White algorithm. I had very smooth pipes (`k_mm = 1e-13`), changed to less smooth (`k_mm = 1e-7`) and got my model working. But my luck didn't last long - the Colebrook-White algorithm didn't converge with any k_mm on pandapipes version 0.10.0 (I was on 0.9.0) 


I'm afraid I can't share the model, but I tried to make a minimal non-working example, but failed - the example worked :) I'm sharing the example for the sake of discussion completeness

<details><summary>Minimal working example</summary>
<p>

![image](https://github.com/e2nIEE/pandapipes/assets/147147326/8ff4d8f6-f080-4cfc-a959-c4aebad4bd68)

Pipe 3-4 is vertical

The example works with both pandapipes 0.9.0 and 0.10.0
```py
import pandapipes as pp

def main():
    net = pp.create_empty_network("net", fluid="lgas")

    pp.create_junctions(net, 5, 1, 300)

    pp.create_ext_grid(net, junction=0, p_bar=5)

    pp.create_pipe_from_parameters(net, from_junction=0, to_junction=1, length_km=0.1, k_mm=1e-15, diameter_m=0.2)
    pp.create_pipe_from_parameters(net, from_junction=1, to_junction=2, length_km=0.1, k_mm=1e-15, diameter_m=0.2)
    pp.create_pipe_from_parameters(net, from_junction=1, to_junction=3, length_km=0.1, k_mm=1e-15, diameter_m=0.2)
    pp.create_pipe_from_parameters(net, from_junction=3, to_junction=4, length_km=0.1, k_mm=1e-15, diameter_m=0.2)

    pp.create_sink(net, junction=2, mdot_kg_per_s=6)

    pp.create_ext_grid(net, junction=4, p_bar=1)

    pp.pipeflow(
        net,
        mode="hydraulics",
        friction_model="colebrook",
        iter=10000,
    )
    print(net.res_junction)
    print(net.res_pipe)

    return net


if __name__ == "__main__":
    net = main()
```

</p>
</details> 


Is the information provided enough to hint me on how to investigate further? If not, I will try to make a non-working example (or find a way to fix my model), but I will only have time for that only in 2 weeks

**Comment by dlohmeier** [comment]
Hi @dad616610 ,

thanks for the example, this always helps. You said that this example works, so I cannot say for sure that any of the following notes could help with the non-working example. What strikes me:

- There is a simulation option "max_iter_colebrook" that you could try to change. The "iter" parameter does not influence the iterations of the Colebrook-White algorithm. Since you said that this is the part that does not converge, increasing this parameter might help.
- You say that your example is working with pandapipes 0.9.0, but not 0.10.0 . I would be really interested in analyzing your example, because we changed quite some internal terms that influence the solution, and we cannot guarantee that everything is working perfectly. We welcome all hints that help us improve the simulation model.
- I also encountered problems with accuracy regarding numba. I did not yet find the solution, but any hints from your side would be welcome here as well. So what you could do is try to set the "use_numba" flag to False. Does that change anything?

I really hope that you can identify what is not working. Otherwise, don't hesitate to provide more insights so that we can help in finding a solution for your problem.
